### PR TITLE
Add optional recording of code coverage

### PR DESCRIPTION
**Pull request by uhafner**
Add a new configuration option `withCoverage` to enable the computation and recording of code coverage results using [JaCoCo maven plugin](https://www.eclemma.org/jacoco/trunk/doc/maven.html) and [Jenkins Code Coverage API plugin](https://github.com/jenkinsci/code-coverage-api-plugin).

See https://ci.jenkins.io/job/Plugins/job/analysis-model/job/PR-484/3/ for an example.

### CODE REVIEW

**Comment by uhafner** [comment]
Replaces #150.

File: vars/buildPlugin.groovy
Comment (timja) [inline vars/buildPlugin.groovy:111]: shouldn't this be done automatically as part of the maven build, using the recommended configuration in the code-coverage-api? 

(skip to jacoco, no anchor unfortunately)
https://github.com/jenkinsci/code-coverage-api-plugin#1-config-coverage-tool-to-generate-reports

File: vars/buildPlugin.groovy
Comment (timja) [inline vars/buildPlugin.groovy:120]: should we enable the `calculateDiffForChangeRequests: true` ?

https://www.jenkins.io/blog/2019/07/21/Jenkins-code-coverage-diff-for-pull-request/#how-to-enable-code-coverage-diff-for-pull-requests

Comment (timja) [reply vars/buildPlugin.groovy:120]: what happens if this is a multi module project?

does the path need to be configurable or should we locate all the jacoco reports?

or YAGNI and enhance later?

File: vars/buildPlugin.groovy
Comment (uhafner) [reply vars/buildPlugin.groovy:111]: Yes, this would be another way to do it. But how do we toggle the execution of Jacoco then? Should we assume that we always record the coverage and remove all toggles?

File: vars/buildPlugin.groovy
Comment (uhafner) [reply vars/buildPlugin.groovy:120]: Yes, adding the diff would make sense, yes. I never used that before but it sounds quite useful.

File: vars/buildPlugin.groovy
Comment (timja) [reply vars/buildPlugin.groovy:111]: is there any reason not to?

Comment (uhafner) [reply vars/buildPlugin.groovy:111]: Normally, I run test twice: once without coverage in order to make sure that the unpatched class files are tested, and a second time with coverage. This duplicates the run time for the tests. But maybe this approach is too academic since in practice I never noticed tests that failed only due the coverage modifications off the class files.

Comment (timja) [reply vars/buildPlugin.groovy:111]: I come from a gradle world where everything like this is just bound to the `check` phase.

It seems odd to me to have to directly invoke plugins, I don't really understand your answer on why you run it without first.

If it adds quite a bit of time to the build I would understand it, although I tried it out on my plugins and it seemed fine.

Comment (uhafner) [reply vars/buildPlugin.groovy:111]: > I don't really understand your answer on why you run it without first.

Because the coverage step changes the class files. I.e. the tests may theoretically fail (or succeed) because of the changed class files. Normally we should test the same executable that we are shipping. (As noted, this might be a too academical approach.(

Comment (uhafner) [reply vars/buildPlugin.groovy]: Do we already have a helper method that detects if files of a given pattern exist?

Comment (timja) [reply vars/buildPlugin.groovy]: Not that I’m aware of

**Comment by timja** [comment]
LGTM pending testing it out on ci.jenkins.io

**Comment by timja** [comment]
I've pushed it as an origin branch to `code-coverage`

Anyone can test this out by adding:

```
@Library('pipeline-library@code-coverage')
```

to their Jenkinsfile.

I'm interested in:

* a pipeline without code coverage
* one with code coverage
* any pom changes required to make it work

**Comment by uhafner** [comment]
I added the following snippet to my parent pom to get jacoco reporting the results:
https://github.com/uhafner/codingstyle/blob/master/pom.xml#L498

We should adapt our parent pom accordingly.

**Comment by uhafner** [comment]
> I've pushed it as an origin branch to `code-coverage`
> 
> Anyone can test this out by adding:
> 
> ```
> @Library('pipeline-library@code-coverage')
> ```
> 
> to their Jenkinsfile.

Do you have a whole example that is working? I tried https://github.com/jenkinsci/analysis-model/pull/495/files but this seems not to compile.

**Comment by timja** [comment]
https://github.com/jenkinsci/analysis-model/pull/495/files#r475479363

**Comment by uhafner** [comment]
Here a first example for a Java library: 
- [PR](https://github.com/jenkinsci/analysis-model/pull/495)
- [GitHub Checks](https://github.com/jenkinsci/analysis-model/pull/495/checks?check_run_id=1021082244)
- [Jenkins CI](https://ci.jenkins.io/job/Plugins/job/analysis-model/job/PR-495/2/)

**Comment by uhafner** [comment]
And here an example of a build that does not produce the coverage results:
https://ci.jenkins.io/job/Plugins/job/forensics-api-plugin/job/serialization-0.7/1/console

```
15:01:00  Publishing Coverage report....
15:01:01  No reports were found
```

**Comment by uhafner** [comment]
BTW: currently the source code visualization is not enabled. Should I enable it for the latest build? (Requires some file system space.)

**Comment by timja** [comment]
> BTW: currently the source code visualization is not enabled. Should I enable it for the latest build? (Requires some file system space.)

May as well enable it

**Comment by uhafner** [comment]
Here is a PR that builds a plugin with coverage: https://github.com/jenkinsci/forensics-api-plugin/pull/203.

**Comment by MarkEWaite** [comment]
> > BTW: currently the source code visualization is not enabled. Should I enable it for the latest build? (Requires some file system space.)
> 
> May as well enable it

I'm `-1` for enabling it at this time because we're 80% full  on the disc that holds the JENKINS_HOME file system on ci.jenkins.io.  Before we enable something that significantly increases disc use on JENKINS_HOME of ci.jenkins.io, we either need to reduce current disc use or add more disc space.

**Comment by uhafner** [comment]
It will not work anyway: https://github.com/jenkinsci/code-coverage-api-plugin/blob/master/src/main/java/io/jenkins/plugins/coverage/targets/CoverageResult.java#L282

The coverage result checks for workspace permission which I do not have.

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
@jenkins-infra/ci-maintainers 🚢 ?

**Comment by MarkEWaite** [comment]
I've freed enough space on ci.jenkins.io that the Jenkins HOME directory is now only 71% full and with some additional time we may get it below 70% full.  I'm OK with shipping this change.

**Pull request merged**
